third_party/terraform/resources/resource_compute_instance.go.erb

```diff
@@ -1389,51 +1423,61 @@ func resourceComputeInstanceUpdate(d *schema.ResourceData, meta interface{}) err
-				}
-				op, err := config.clientComputeBeta.Instances.UpdateNetworkInterface(project, zone, instance.Name, networkName, ni).Do()
-				if err != nil {
-					return errwrap.Wrapf("Error removing alias_ip_range: {{err}}", err)
-				}
-				opErr := computeOperationWaitTime(config, op, project, "updating alias ip ranges", d.Timeout(schema.TimeoutUpdate))
-				if opErr != nil {
-					return opErr
-				}
-				rereadFingerprint = true
+			//re-read fingerprint
+			instance, err = config.clientComputeBeta.Instances.Get(project, zone, instance.Name).Do()
+			if err != nil {
+				return err
 			}
+			instNetworkInterface = instance.NetworkInterfaces[i]
+		}
 
-			ranges := d.Get(prefix + ".alias_ip_range").([]interface{})
-			if len(ranges) > 0 {
-				if rereadFingerprint {
+		if !updateDuringStop {
+			if d.HasChange(prefix + ".alias_ip_range") {
+				// Alias IP ranges cannot be updated; they must be removed and then added.
+				// unless you are changing subnetwork/network
+				if len(instNetworkInterface.AliasIpRanges) > 0 {
+					ni := &computeBeta.NetworkInterface{
+						Fingerprint:     instNetworkInterface.Fingerprint,
+						ForceSendFields: []string{"AliasIpRanges"},
+					}
+					op, err := config.clientComputeBeta.Instances.UpdateNetworkInterface(project, zone, instance.Name, networkName, ni).Do()
+					if err != nil {
+						return errwrap.Wrapf("Error removing alias_ip_range: {{err}}", err)
+					}
+					opErr := computeOperationWaitTime(config, op, project, "updating alias ip ranges", d.Timeout(schema.TimeoutUpdate))
+					if opErr != nil {
+						return opErr
+					}
+					//re-read fingerprint
 					instance, err = config.clientComputeBeta.Instances.Get(project, zone, instance.Name).Do()
 					if err != nil {
 						return err
 					}
 					instNetworkInterface = instance.NetworkInterfaces[i]
 				}
-				ni := &computeBeta.NetworkInterface{
-					AliasIpRanges: expandAliasIpRanges(ranges),
-					Fingerprint:   instNetworkInterface.Fingerprint,
-				}
-				op, err := config.clientComputeBeta.Instances.UpdateNetworkInterface(project, zone, instance.Name, networkName, ni).Do()
-				if err != nil {
-					return errwrap.Wrapf("Error adding alias_ip_range: {{err}}", err)
-				}
-				opErr := computeOperationWaitTime(config, op, project, "updating alias ip ranges", d.Timeout(schema.TimeoutUpdate))
-				if opErr != nil {
-					return opErr
-				}
 			}
 		}
+
+		// Setting NetworkIP to empty and AccessConfigs to nil.
+		// This will opt them out from being modified in the patch call.
+		// They cannot be changed by UpdateNetworkInterface
+		networkInterface.NetworkIP = ""
+		networkInterface.AccessConfigs = nil
+		networkInterface.Fingerprint = instNetworkInterface.Fingerprint
+		updateCall := config.clientComputeBeta.Instances.UpdateNetworkInterface(project, zone, instance.Name, networkName, networkInterface).Do
+		if !updateDuringStop && d.HasChange(prefix+".alias_ip_range") {
+			op, err := updateCall()
```

Comment:
I think it'll be clearer to have the one if block that does the remove/add for alias ip ranges. It took me a few reads before I realized that this was the "add" component of the remove/add that's called out in line 1437, since logically that section of code appeared to have ended.

Reply:
Will update